BUG: data as observed in RV

### Describe the issue:

Example:

https://github.com/pymc-labs/pymc-marketing/issues/1383#issuecomment-2595717439

### Reproduceable code example:

```python
import numpy as np
import pymc as pm

seed = sum(map(ord, "Scaling the likelihood depended variables doesn't work in PyMC"))
rng = np.random.default_rng(seed)

true_mu = 100
true_sigma = 30

n_obs = 10
coords = {
    "date": np.arange(n_obs),
}

dist = pm.Normal.dist(mu=true_mu, sigma=true_sigma, shape=n_obs)
data = pm.draw(dist, random_seed=rng)

scaling = data.max()

with pm.Model(coords=coords) as model:
    mu = pm.Normal("mu")
    sigma = pm.HalfNormal("sigma")

    target = pm.Data("target", data, dims="date")
    scaled_target = target / scaling

    pm.Normal("observed", mu=mu, sigma=sigma, observed=scaled_target)
```

### Error message

```
            if not is_valid_observed(observed):
>               raise TypeError(
                    "Variables that depend on other nodes cannot be used for observed data."
                    f"The data variable was: {observed}"
                )
E               TypeError: Variables that depend on other nodes cannot be used for observed data.The data variable was: True_div.0
```

### PyMC version information:

5.20.0

### Context for the issue:

Would like to define my scaling within the model definition

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -190,7 +190,7 @@ def Minibatch(variable: TensorVariable, *variables: TensorVariable, batch_size: 
     for i, v in enumerate(tensors):
         if not is_valid_observed(v):
             raise ValueError(
-                f"{i}: {v} is not valid for Minibatch, only constants or constants.astype(dtype) are allowed"
+                f"{i}: {v} is not valid for Minibatch, only non-random variables are allowed"
             )
 
     upper = tensors[0].shape[0]

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -33,6 +33,7 @@ from pytensor.raise_op import Assert
 from pytensor.scalar import Cast
 from pytensor.tensor.elemwise import Elemwise
 from pytensor.tensor.random.basic import IntegersRV
+from pytensor.tensor.random.var import RandomGeneratorSharedVariable
 from pytensor.tensor.type import TensorType
 from pytensor.tensor.variable import TensorConstant, TensorVariable
 
@@ -148,6 +149,19 @@ class MinibatchOp(OpFromGraph):
         return "Minibatch"
 
 
+def first_inputs(r):
+    if not r.owner:
+        return
+
+    first_input = r.owner.inputs[0]
+    yield first_input
+    yield from first_inputs(first_input)
+
+
+def has_random_ancestor(r):
+    return any(isinstance(i, RandomGeneratorSharedVariable) for i in first_inputs(r))
+
+
 def is_valid_observed(v) -> bool:
     if not isinstance(v, Variable):
         # Non-symbolic constant
@@ -165,6 +179,7 @@ def is_valid_observed(v) -> bool:
             and isinstance(v.owner.op.scalar_op, Cast)
             and is_valid_observed(v.owner.inputs[0])
         )
+        or not has_random_ancestor(v)
         # Or Minibatch
         or (
             isinstance(v.owner.op, MinibatchOp)

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -30,15 +30,13 @@ from pytensor.compile.builders import OpFromGraph
 from pytensor.compile.sharedvalue import SharedVariable
 from pytensor.graph.basic import Variable
 from pytensor.raise_op import Assert
-from pytensor.scalar import Cast
-from pytensor.tensor.elemwise import Elemwise
 from pytensor.tensor.random.basic import IntegersRV
-from pytensor.tensor.random.var import RandomGeneratorSharedVariable
 from pytensor.tensor.type import TensorType
 from pytensor.tensor.variable import TensorConstant, TensorVariable
 
 import pymc as pm
 
+from pymc.logprob.utils import rvs_in_graph
 from pymc.pytensorf import GeneratorOp, convert_data, smarttypeX
 from pymc.vartypes import isgenerator
 
@@ -149,22 +147,8 @@ class MinibatchOp(OpFromGraph):
         return "Minibatch"
 
 
-def first_inputs(r):
-    if not r.owner:
-        return
-
-    inputs = r.owner.inputs
-
-    if not inputs:
-        return
-
-    first_input = inputs[0]
-    yield first_input
-    yield from first_inputs(first_input)
-
-
 def has_random_ancestor(r):
-    return any(isinstance(i, RandomGeneratorSharedVariable) for i in first_inputs(r))
+    return rvs_in_graph(r) != set()
 
 
 def is_valid_observed(v) -> bool:
@@ -177,14 +161,7 @@ def is_valid_observed(v) -> bool:
         return True
 
     return (
-        # The only PyTensor operation we allow on observed data is type casting
-        # Although we could allow for any graph that does not depend on other RVs
-        (
-            isinstance(v.owner.op, Elemwise)
-            and isinstance(v.owner.op.scalar_op, Cast)
-            and is_valid_observed(v.owner.inputs[0])
-        )
-        or not has_random_ancestor(v)
+        not has_random_ancestor(v)
         # Or Minibatch
         or (
             isinstance(v.owner.op, MinibatchOp)

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -153,7 +153,12 @@ def first_inputs(r):
     if not r.owner:
         return
 
-    first_input = r.owner.inputs[0]
+    inputs = r.owner.inputs
+
+    if not inputs:
+        return
+
+    first_input = inputs[0]
     yield first_input
     yield from first_inputs(first_input)
 

--- a/pymc/data.py
+++ b/pymc/data.py
@@ -147,10 +147,6 @@ class MinibatchOp(OpFromGraph):
         return "Minibatch"
 
 
-def has_random_ancestor(r):
-    return rvs_in_graph(r) != set()
-
-
 def is_valid_observed(v) -> bool:
     if not isinstance(v, Variable):
         # Non-symbolic constant
@@ -161,7 +157,7 @@ def is_valid_observed(v) -> bool:
         return True
 
     return (
-        not has_random_ancestor(v)
+        not rvs_in_graph(v)
         # Or Minibatch
         or (
             isinstance(v.owner.op, MinibatchOp)

--- a/pymc/pytensorf.py
+++ b/pymc/pytensorf.py
@@ -177,9 +177,9 @@ def extract_obs_data(x: TensorVariable) -> np.ndarray:
             mask[mask_idx] = 1
             return np.ma.MaskedArray(array_data, mask)
 
-    from pymc.data import has_random_ancestor
+    from pymc.logprob.utils import rvs_in_graph
 
-    if not has_random_ancestor(x):
+    if not rvs_in_graph(x):
         cheap_eval_mode = Mode(linker="py", optimizer=None)
         return x.eval(mode=cheap_eval_mode)
 

--- a/pymc/pytensorf.py
+++ b/pymc/pytensorf.py
@@ -179,7 +179,7 @@ def extract_obs_data(x: TensorVariable) -> np.ndarray:
 
     from pymc.logprob.utils import rvs_in_graph
 
-    if not rvs_in_graph(x):
+    if not inputvars(x) and not rvs_in_graph(x):
         cheap_eval_mode = Mode(linker="py", optimizer=None)
         return x.eval(mode=cheap_eval_mode)
 

--- a/pymc/pytensorf.py
+++ b/pymc/pytensorf.py
@@ -180,7 +180,8 @@ def extract_obs_data(x: TensorVariable) -> np.ndarray:
     from pymc.data import has_random_ancestor
 
     if not has_random_ancestor(x):
-        return x.eval()
+        cheap_eval_mode = Mode(linker="py", optimizer=None)
+        return x.eval(mode=cheap_eval_mode)
 
     raise TypeError(f"Data cannot be extracted from {x}")
 

--- a/pymc/pytensorf.py
+++ b/pymc/pytensorf.py
@@ -177,6 +177,11 @@ def extract_obs_data(x: TensorVariable) -> np.ndarray:
             mask[mask_idx] = 1
             return np.ma.MaskedArray(array_data, mask)
 
+    from pymc.data import has_random_ancestor
+
+    if not has_random_ancestor(x):
+        return x.eval()
+
     raise TypeError(f"Data cannot be extracted from {x}")
 
 

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -547,3 +547,10 @@ def test_scaling_data_works_in_likelihood() -> None:
         scaled_target = target / scale
         mu = pm.Normal("mu", mu=0, sigma=1)
         pm.Normal("x", mu=mu, sigma=1, observed=scaled_target)
+
+        idata = pm.sample(10, chains=1, tune=100)
+
+    np.testing.assert_allclose(
+        idata.observed_data["x"].values,
+        data / scale,
+    )

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -628,7 +628,7 @@ class TestMinibatch:
 def test_scaling_data_works_in_likelihood() -> None:
     data = np.array([10, 11, 12, 13, 14, 15])
 
-    with pm.Model() as model:
+    with pm.Model():
         target = pm.Data("target", data)
         scale = 12
         scaled_target = target / scale

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -602,11 +602,11 @@ class TestMinibatch:
         mb = pm.Minibatch(pt.as_tensor(self.data).astype(int), batch_size=20)
         assert isinstance(mb.owner.op, MinibatchOp)
 
-        with pytest.raises(ValueError, match="not valid for Minibatch"):
-            pm.Minibatch(pt.as_tensor(self.data) * 2, batch_size=20)
+        mb = pm.Minibatch(pt.as_tensor(self.data) * 2, batch_size=20)
+        assert isinstance(mb.owner.op, MinibatchOp)
 
-        with pytest.raises(ValueError, match="not valid for Minibatch"):
-            pm.Minibatch(self.data, pt.as_tensor(self.data) * 2, batch_size=20)
+        for mb in pm.Minibatch(self.data, pt.as_tensor(self.data) * 2, batch_size=20):
+            assert isinstance(mb.owner.op, MinibatchOp)
 
     def test_assert(self):
         d1, d2 = pm.Minibatch(self.data, self.data[::2], batch_size=20)

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -608,6 +608,12 @@ class TestMinibatch:
         for mb in pm.Minibatch(self.data, pt.as_tensor(self.data) * 2, batch_size=20):
             assert isinstance(mb.owner.op, MinibatchOp)
 
+    def test_not_allowed(self):
+        data = pt.random.normal(loc=self.data, scale=1)
+
+        with pytest.raises(ValueError):
+            pm.Minibatch(data, batch_size=20)
+
     def test_assert(self):
         d1, d2 = pm.Minibatch(self.data, self.data[::2], batch_size=20)
         with pytest.raises(

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -623,3 +623,14 @@ class TestMinibatch:
         [draw_mA, draw_mB] = pm.draw([mA, mB])
         assert draw_mA.shape == (10,)
         np.testing.assert_allclose(draw_mA, -draw_mB)
+
+
+def test_scaling_data_works_in_likelihood() -> None:
+    data = np.array([10, 11, 12, 13, 14, 15])
+
+    with pm.Model() as model:
+        target = pm.Data("target", data)
+        scale = 12
+        scaled_target = target / scale
+        mu = pm.Normal("mu", mu=0, sigma=1)
+        pm.Normal("x", mu=mu, sigma=1, observed=scaled_target)

--- a/tests/test_pytensorf.py
+++ b/tests/test_pytensorf.py
@@ -207,6 +207,13 @@ class TestExtractObsData:
         assert isinstance(res, np.ndarray)
         np.testing.assert_array_equal(res, np.array([4, 7, 10]))
 
+    def test_pytensor_operations_raises(self):
+        x = pt.scalar("x")
+        target = 1 + 3 * x
+
+        with pytest.raises(TypeError, match="Data cannot be extracted from"):
+            extract_obs_data(target)
+
 
 @pytest.mark.parametrize("input_dtype", ["int32", "int64", "float32", "float64"])
 def test_convert_data(input_dtype):

--- a/tests/test_pytensorf.py
+++ b/tests/test_pytensorf.py
@@ -199,6 +199,14 @@ class TestExtractObsData:
         assert isinstance(res, np.ndarray)
         np.testing.assert_array_equal(res, y)
 
+    def test_pytensor_operations(self):
+        x = np.array([1, 2, 3])
+        target = 1 + 3 * pt.as_tensor_variable(x)
+
+        res = extract_obs_data(target)
+        assert isinstance(res, np.ndarray)
+        np.testing.assert_array_equal(res, np.array([4, 7, 10]))
+
 
 @pytest.mark.parametrize("input_dtype", ["int32", "int64", "float32", "float64"])
 def test_convert_data(input_dtype):

